### PR TITLE
feat: Support repositories and permissions in GitHub generator

### PR DESCRIPTION
**Pull request by konradasb**
## Problem Statement

We want to limit the scope of the GitHub access token that the GitHub generator creates. It's possible to do so [according to GitHub API](https://docs.github.com/en/rest/apps/apps?apiVersion=2022-11-28#create-an-installation-access-token-for-an-app)

## Checklist

- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [x] My changes have reasonable test coverage
- [x] All tests pass with `make test`
- [x] I ensured my PR is ready for review with `make reviewable`

### CODE REVIEW

File: pkg/generator/github/github.go [hunk 1]
```diff
@@ -80,8 +83,21 @@ func (g *Generator) generate(
...
+		body["repositories"] = gh.Repositories
+	}
+	bodyBytes, err := json.Marshal(body)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %w", err)
+	}
+
 	// Github api expects POST request
-	req, err := http.NewRequestWithContext(ctx, http.MethodPost, gh.URL, http.NoBody)
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, gh.URL, bytes.NewReader(bodyBytes))
```
Comment (Skarlso) [inline pkg/generator/github/github.go:100]: This is slightly different as opposed to not sending a body at all. I would say, make the body nil and only created one if the things are set.

so something like:

```go
body := http.NoBody
...
if bla
body, err = bytes.NewReader(bodyBytes)
...


...
req, err := http.NewRequestWithContext(ctx, http.MethodPost, gh.URL, body)
```

File: pkg/generator/github/github.go [hunk 2]
```diff
@@ -120,7 +136,8 @@ func newGHClient(ctx context.Context, k client.Client, n string, hc *http.Client
 	if err != nil {
 		return nil, fmt.Errorf(errParseSpec, err)
 	}
-	gh := &Github{Kube: k, Namespace: n, HTTP: hc}
+	gh := &Github{Kube: k, Namespace: n, HTTP: hc,
```
Comment (Skarlso) [inline pkg/generator/github/github.go:139]: Please fmt this line. :D

File: pkg/generator/github/github.go [hunk 3]
```diff
@@ -84,20 +85,28 @@ func (g *Generator) generate(
...
+	if len(payload) > 0 {
+		bodyBytes, err := json.Marshal(payload)
+		if err != nil {
+			return nil, fmt.Errorf("error marshalling payload: %w", err)
+		}
+
+		body = bytes.NewReader(bodyBytes)
+	} else {
+		body = http.NoBody
 	}
```
Comment (Skarlso) [inline pkg/generator/github/github.go:106]:
```suggestion
	body := http.NoBody
	if len(payload) > 0 {
		bodyBytes, err := json.Marshal(payload)
		if err != nil {
			return nil, fmt.Errorf("error marshalling payload: %w", err)
		}

		body = bytes.NewReader(bodyBytes)
	}
```

No need for the else.

Comment (konradasb) [reply pkg/generator/github/github.go:106]: sorry, 
```
cannot use bytes.NewReader(bodyBytes) (value of type *bytes.Reader) as http.noBody value in assignment
```

Comment (Skarlso) [reply pkg/generator/github/github.go:106]: Ah yes. Use this:
```go
	var body io.Reader = http.NoBody
	if len(payload) > 0 {
		bodyBytes, err := json.Marshal(payload)
		if err != nil {
			return nil, fmt.Errorf("error marshalling payload: %w", err)
		}

		body = bytes.NewReader(bodyBytes)
	}
```

Comment (konradasb) [reply pkg/generator/github/github.go]: This worked, learned something new, thanks!

**Comment by Skarlso** [comment]
@konradasb You still have some linter issues. :)

**Comment by konradasb** [comment]
Should be OK now

**Review by Skarlso — APPROVED** [review]

**Pull request merged**